step3-자동차 경주

kimdh-hi opened the pull request:
안녕하세요 리뷰어님! 
자동차 경주 리뷰드립니다.

이번에도 많이 부족하지만 잘 부탁드립니다!!🙇‍♂️

kimdh-hi committed:
docs(racing): 요구사항 정의

kimdh-hi committed:
feat(racing): step3 racing game 구현

kimdh-hi committed:
docs(racing) .

ohgillwhan commented on src/main/kotlin/racing/Car.kt:
```diff
@@ -0,0 +1,18 @@
+
+class Car(
+    var distance: Int = INITIAL_DISTANCE,
+    private val forwardCondition: ForwardCondition
```
전략패턴 좋네요~

ohgillwhan commented on src/main/kotlin/racing/Car.kt:
```diff
@@ -0,0 +1,18 @@
+package racing
+
+class Car(
+    var distance: Int = INITIAL_DISTANCE,
```
`var`로 지정하고 `private`으로 접근제어자를 하지 않으면, 외부에서도 수정이 가능해요. `private`으로 바꿀 수 있을까요?

ohgillwhan commented on src/main/kotlin/racing/Car.kt:
```diff
@@ -0,0 +1,18 @@
+    private val forwardCondition: ForwardCondition
+) {
+
+    companion object {
+        const val INITIAL_DISTANCE = 1
+    }
+
+    fun move() {
+        val forwardCondition = forwardCondition.getForwardCondition()
+        if (forwardCondition) {
+            ++distance
+        }
+    }
```
https://kotlinlang.org/docs/coding-conventions.html#class-layout

선언 순서를 확인해볼까요?

ohgillwhan commented on src/main/kotlin/racing/Cars.kt:
```diff
@@ -0,0 +1,19 @@
+        }
+    }
+
+    fun racing() {
```
동사로 정의해야 하기 떄문에 `race` 는 어떨까요?

ohgillwhan commented on src/main/kotlin/racing/Cars.kt:
```diff
@@ -0,0 +1,19 @@
+package racing
+
+class Cars(
+    val cars: List<Car>
```
이 부분도 충분히 `private`으로 가능하지 않을까요? 함수의 리턴값을 잘 활용하면 될거에요

ohgillwhan commented on src/main/kotlin/racing/InputView.kt:
```diff
@@ -0,0 +1,16 @@
+
+    private const val ERROR_MESSAGE = "0 이상의 정수를 입력해주세요."
+
+    fun input(): Pair<Int, Int> {
```
`Pair`에 같은 타입이 2개가 오면 어느게 무슨 값인지 햇갈려요. 이럴 때는 `DTO` 를 만드는게 더 좋을수 있어 보여요.

ohgillwhan commented on src/main/kotlin/racing/OutputView.kt:
```diff
@@ -0,0 +1,19 @@
+    private fun output(racingPhasedRecords: List<RacingPhasedRecord>) {
+        racingPhasedRecords.forEach { phaseRecord ->
+            phaseRecord.records.forEach { record ->
+                repeat(record) { print("-") }
+                println()
```
`println("-".repeat(record))`도 가능할거 같아요 👍

ohgillwhan commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,22 @@
+
+    fun start() {
+        repeat(attemptCount) {
+            cars.racing()
```
이 부분에서 리턴값을 잘 활용하면 이 클래스의 코드가 전체적으로 조금 더 깔끔할 수 있겠네요!

ohgillwhan commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,22 @@
+    val racingPhasedRecords: MutableList<RacingPhasedRecord> = mutableListOf()
+
+    fun start() {
+        repeat(attemptCount) {
```
`(0..ateemptCount).map{}` 을 사용하면 `racingPhasedRecords`는 `mutable` 하지 않아도 될거에요!

kimdh-hi replied:
지난 미션에도 리뷰주셨던 부분인데 신경쓰지 못했네요.. 리뷰 감사합니다!

ohgillwhan reviewed (changes requested):
코드 잘 봤습니다! 전체적으로 깔끔 하네요!
리뷰 확인 부탁드립니다.

kimdh-hi committed:
refactor(racing): 자동차 경주 코드리뷰 반영

ohgillwhan commented on src/main/kotlin/racing/Car.kt:
```diff
@@ -0,0 +1,20 @@
+        }
+    }
+
+    fun getDistance() = distance
```
`getter`는 최대한 없애보아요. `move()`의 리턴을 활용해보면 자연스레 잘 될거같아요~

ohgillwhan commented on src/main/kotlin/racing/Car.kt:
```diff
@@ -0,0 +1,20 @@
+    fun getDistance() = distance
+
+    companion object {
+        const val INITIAL_DISTANCE = 1
```
💯

ohgillwhan commented on src/main/kotlin/racing/Cars.kt:
```diff
@@ -0,0 +1,20 @@
+        }
+    }
+
+    fun race(): List<Car> {
+        cars.forEach { car -> car.move() }
+        return cars
+    }
```
https://kotlinlang.org/docs/coding-conventions.html#class-layout 확인해볼까요?

ohgillwhan replied:
자동차 전체를 반환하게 된다면, 외부에서 자동차에서 제공하는 함수를 쉽게 호출할 수 있게되면서 `move()`를 호출할 수 있는 취약점이 생길 것 같아요. 
`move()`의 반환값을 적절하게 만들고, 그 값들을 모아서 반환해볼까요? 어떤 장단점이 있을까요?

kimdh-hi replied:
List<Car> 를 반환한다면 List<Car> 프로퍼티를 private 으로 선언한 의미가 없어질수도 있겠네요.
말씀주신대로 Car 가 불필요하게 외부에 공개되는 문제가 생길 수 있을 것 같습니다.

따로 결과값을 위한 클래스를 정의해서 사용해야지 Car 를 공개하지 않기로 한 목적을 제대로 지킬 수 있습니다.

ohgillwhan replied:
실무에서는 클린코드도 중요하지만, 생산성 또한 중요하기 때문에 어느정도의 트레이드 오프로 `private`를 쓰지 않을 수 있습니다! 미션을 하면서 최대한 클린코드, 객체지향을 중심적으로 작성해보면 좋겠습니다!

ohgillwhan commented on src/main/kotlin/racing/OutputView.kt:
```diff
@@ -0,0 +1,14 @@
+
+object OutputView {
+
+    fun output(racingPhasedRecords: List<RacingPhasedRecord>) {
```
출력하다라는 `print` 라는 이름도 괜찮아 보여요.

ohgillwhan commented on src/main/kotlin/racing/InputView.kt:
```diff
@@ -0,0 +1,16 @@
+
+    private const val ERROR_MESSAGE = "0 이상의 정수를 입력해주세요."
+
+    fun input(): RacingInputDto {
```
읽다의 `read`도 좋은거 같아요.

ohgillwhan commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,13 @@
+    private val attemptCount: Int
+) {
+
+    fun start() = (0 until attemptCount).map {
+        cars.race().map { car ->
+            car.getDistance()
+        }
+    }.map { RacingPhasedRecord(it) }
```
코틀린에서 `=` 를 통해 함수를 정의할 수 있는데, 코드가 조금 복잡해지면 `{}`를 통해 반환하는게 더 보기 편해집니다!

kimdh-hi replied:
당연히 명확한 기준을 세우기 어려운 부분이겠지만 `=` 를 사용하는, 복잡하지 않은 함수의 기준이 있으신가요?

ohgillwhan replied:
거의 `=`를 사용하지 않고 있습니다. 사용한다면 간단한 사칙연산이나, 로직이 들어가지 않는 `.size()` 정도로만 하고 있어요.

ohgillwhan commented on src/main/kotlin/racing/RandomNumberForwardCondition.kt:
```diff
@@ -0,0 +1,17 @@
+        private val random = Random()
+    }
+
+    override fun getForwardCondition(): Boolean {
+        return random.nextInt(MAX_FORWARD_CONDITION) >= MIN_FORWARD_CONDITION
+    }
```
`convention` 확인 부탁드립니다~

ohgillwhan commented on src/main/kotlin/racing/RandomNumberForwardCondition.kt:
```diff
@@ -0,0 +1,17 @@
+package racing
+
+import java.util.Random
```
`kotlin`의 `random`으로 해볼까요?

ohgillwhan reviewed (changes requested):
코드 잘 봤습니다! 조금 더 고민해볼만한 리뷰 남겨봤습니다! 확인 부탁드려요~

kimdh-hi committed:
refactor(racing): 자동차 경주 코드리뷰 반영

- RacingPhaseRecord, RacingPhaseRecords 추가
- Cars.race 의 반환값으로 List<Car> 를 반환하는 부분 별도 결과값을 담은 객체가 반환되도록 수정

kimdh-hi committed:
refactor(racing): RacingPhaseRecord Car 에서 반환하도록 수정

ohgillwhan commented on src/main/kotlin/racing/Cars.kt:
```diff
@@ -0,0 +1,17 @@
+        }
+    }
+
+    fun race(): List<RacingPhaseRecord> = cars.map { it.move() }
```
이 코드도 `companion object` 위로 올라와야겠네요~

ohgillwhan commented on src/test/kotlin/racing/CarsTest.kt:
```diff
@@ -0,0 +1,19 @@
+        val cars = Cars.of(carCount, RandomNumberForwardCondition())
+
+        // then
+        assertThat(cars.race().size).isEqualTo(carCount)
```
사이드 이펙이 없는 `cars.size()` 같은 메서드를 만들어도 괜찮을것 같네요!

ohgillwhan reviewed (approved):
코드 잘 작성해주셨습니다!
`UI`와 `domain`을 패키지를 나눠보셔도 좋을것 같네요.

다음 `step`도 진행 부탁드립니다.

진행하면서 필요시 아래의 글도 참고해보세요

https://kotlinlang.org/docs/properties.html#backing-properties
